### PR TITLE
Fix bugs in the constraint solver for `And`

### PR DESCRIPTION
**Pull request by yihozhang**

### CODE REVIEW

File: tests/type-constraints-tests.egg [hunk 1]
```diff
@@ -0,0 +1,10 @@
+(datatype Operand)
+(sort VecOperandBase (Vec Operand))
+(datatype VecOperand (VO VecOperandBase))
```
Comment (saulshanabrook) [inline tests/type-constraints-tests.egg:3]: Sorry I think I missed it maybe in an earlier PR, what does the `V0` do?

Comment (yihozhang) [reply tests/type-constraints-tests.egg:3]: Ah it's a wrapper trick we used in eggcc that allows us to union two vectors. It's very specific to our application.

Comment (saulshanabrook) [reply tests/type-constraints-tests.egg:3]: Ah sorry, I was just misreading the datatype constructor as a sort constructor and thought that `V0` was a presort. Nevermind!

**Review by oflatt — APPROVED** [review]
Looks good to me

**Pull request merged**